### DISCUSSION TITLE
JAVA - 1 - 호현, 일구

### DISCUSSION THREAD

**Discussion by sikutisa** [Shortudy-10th/tech-for-developer#18] [질의 응답]
## Java 1일차 질문 목록
* 자바 소스 파일의 컴파일 & 실행 과정에 대해 설명해주세요.
* Call By Value / Call By Reference에 대해 설명해주세요.
* 자바의 자료형에 대해 아는 대로 설명해주세요.
* Java Collection에 대해서 설명해주세요.
* JAVA에는 접근 제어자가 4가지 있습니다. 각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.

**Comment by sikutisa** [comment]
### ❓️자바 소스 파일의 컴파일 & 실행 과정에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>
.java 작성
컴파일러(Java Compiler)가 소스파일을 컴파일해 .class라는 바이트코드 파일 생성.

컴파일된 바이트코드를 JVM의 클래스 로더(Class Loader)에게 전달.
클래스 로더는 동적로딩(Dynamic Loading)을 통해 필요한 클래스들을 로딩 및 링크하여 런타임 데이터 영역(Runtime Data area), 즉 JVM의 메모리에 올림.
실행엔진(Execution Engine)은 JVM 메모리에 올라온 바이트 코드들을 명령어 단위로 하나씩 가져와서 실행.
</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️클래스 로더 세부 동작 방식

<br>

### 💬 답변
<blockquote>
로드 : 클래스 파일을 가져와서 JVM의 메모리에 로드

검증 : 자바 언어 명세(Java Language Specification) 및 JVM 명세에 명시된 대로 구성되어 있는지 검사
준비 : 클래스가 필요로 하는 메모리를 할당 (필드, 메서드, 인터페이스 등등)
분석 : 클래스의 상수 풀 내 모든 심볼릭 레퍼런스를 다이렉트 레퍼런스로 변경
초기화 : 클래스 변수들을 적절한 값으로 초기화(static 필드)
</blockquote>

<br>

**Comment by sikutisa** [comment]
### ❓️Call By Value / Call By Reference에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>
값을 parameter로 전달할 때 값이 전달되는 방식

* Call By Value : parameter의 값을 복사해 생성한 새로운 data를 전달하는 개념, Java에서는 primitive type과 reference type 모두  Call by Value 방식으로만 전달 가능

* Call By Reference : parameter 자체를 전달하는 개념, Java에서는 reference type parameter을 전달할 때 개념만 이용하고 실제로는 Call by Value로 전달
</blockquote>

<br>

**Comment by sikutisa** [comment]
### ❓️자바의 자료형에 대해 아는 대로 설명해주세요.

<br>

### 💬 답변
<blockquote>

크게 primitive와 reference로 구분
primitive: byte(1), char(2), short(2), int(4), long(8), float(4), double(8), boolean(1)
reference: 각종 클래스 인스턴스와 배열

</blockquote>

<br>

**Reply by Hohyun-Kim** [reply]
### ❓️primitive type과 reference type의 자료들은 선언 및 정의할 때 어떤 차이가 있나요?

<br>

### 💬 답변
<blockquote>

* 선언의 가장 큰 차이점은 new 키워드 사용여부이다. 
* primitive 타입은 대입연산자만을 사용해서 선언할 수 있다.
   * e.g. int a = 10;
* reference type은 new 키워드와 생성자를 사용해야 한다.
   * e.g. StringBuilder builder = new StringBuilder();
   * null로 초기화할 수도 있다.
      * e.g. StringBuilder builder = null;
   * 배열 역시 referencetype이므로 new를 사용한다.
      * e.g. int[] arr = new int[10];
   * String은 예외적으로 대입연산자만으로 초기화할 수 있다.
      * e.g. String str = "string"


</blockquote>

<br>

**Comment by sikutisa** [comment]
### ❓️Java Collection에 대해서 설명해주세요.

<br>

### 💬 답변
<blockquote>
자바에서 데이터를 효과적으로 처리하고 다루기 위해 제공하는 클래스의 집합.

Collection을 상속받은 List, Set과 Map으로 구성
List: 순서가 있는 데이터의 집합, 중복 허용
Set: 순서가 없는 데이터의 집합, 중복 불가
Map: Key-Value 쌍으로 구성된 데이터의 집합. Key의 중복 불가
</blockquote>

<br>

**Comment by sikutisa** [comment]
### ❓️JAVA에는 접근 제어자가 4가지 있습니다. 각각의 이름을 말씀해 주시고, 각각 제어자의 차이점을 설명해 주세요.

<br>

### 💬 답변
<blockquote>
private: class 외부에서 접근이 불가능한 제어자

default: private의 스코프 + 같은 패키지에 속한 클래스에서는 접근 가능한 제어자
protected: default의 스코프 + 다른 패키지라도 이 클래스를 상속한 클래스에서는 접근 가능한 제어자
public: 접근 제한이 없는 제어자
</blockquote>

<br>
